presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/README.md

```diff
@@ -0,0 +1,169 @@
+# Presto planner and optimizer tests
+
+## Quickstart
+
+### The basics
+
+If you want to see examples of basic use of the plan testing framework, look at
+`TestPlanFramework`. If you'd like to try to begin writing new tests, look at
+`PlanMatchPattern` to see what sorts of patterns are already well-supported by
+the framework.
+
+If you're dealing with `Expressions`, you'll probably want to take a look at
+`ExpressionVerifier` as well.
+
+### Testing planner code
+
+Planner code should be tested by comparing the unoptimized plan to a set of
+expectations. For various reasons related to how `Symbols` are handled in
+unoptimized plans, this is frequently impossible. The minimum set of optimizers
+that needs to be applied for the plan matching framework to functin is likely
+to be some subset of the following:
+
+- `UnaliasSymbolReferences`
+- `PruneIdentityProjections`
+- `PruneUnreferencedOutputs`
+
+`TestQuantifiedComparison` illustrates unit testing planner code.
+
+### Testing optimizer code
+
+Optimizer code will generally need to be subjected to two types of tests:
+
+1. Unit tests on the minimum set of optimizers needed for the optimizer to
+   function.
+2. A test using the full set of optimizers that ensures that the optimizer
+   continues to function after the introduction of new optimizers or reordering
+   of existing optimizers.
+
+`TestMergeWindows` illustrates both types of tests and goes further into the
+rationale for both types of tests.
+
+### Supporting new PlanNodes
+
+There are several things you might want to verify in a `PlanNode`:
+
+1. The contents of the PlanNode itself (e.g. equi criteria in a `JoinNode`)
+2. The contents of some `Symbol`-producing member of the `PlanNode` (e.g.
+   `FunctionCalls` in a `WindowNode`)
+
+These should be handled separately.
+
+#### Fields that do not produce Symbols
+
+Fields that do not produce `Symbols` should be verified in a new `Matcher` that
+is specific to the type of the `PlanNode` being verified. As much as possible,
+the `Matcher` should use `.equals()` methods belonging to the objects being
+compared rather than implementing the comparison itself.
+
+See `JoinMatcher` for an example of comparing non-symbol producing fields.
+
+#### Fields that produce Symbols
```

Comment:
What do you mean by fields? I thought it was fields in the class, but below in your examples you map all of TableScanNode to ColumnReference.

To clarify this, maybe say "PlanNode fields..." in the title and have something like " 'FunctionCall' in 'AggregationNode'" in the table